rust/primary/src/messages.rs

```diff
@@ -1,189 +1,255 @@
     }
 }
 
-/// This is the data structure the receive re-computes and stores.
-#[derive(Debug, Clone, Serialize, Deserialize)]
-pub struct SignedBlockHeader {
-    /// The blockheader data.
-    pub data: Vec<u8>,
-    /// The digest of the blockheader.
-    pub digest: Digest,
-    /// A signature on the digest of the blockheader.
-    signature: Signature,
+impl fmt::Debug for Header {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(
+            f,
+            "{}: B{}({}, {})",
+            self.id,
+            self.round,
+            self.author,
+            self.payload.keys().map(|x| x.size()).sum::<usize>(),
+        )
+    }
 }
 
-impl SignedBlockHeader {
-    pub async fn new(
-        header: BlockHeader,
-        signing_channel: &mut Sender<(Digest, oneshot::Sender<Signature>)>,
-    ) -> Result<Self, DagError> {
-        let data: Vec<u8> = bincode::serialize(&header)?;
-
-        let mut h: Sha512 = Sha512::new();
-        let mut hash: [u8; 64] = [0u8; 64];
-        let mut digest: [u8; 32] = [0u8; 32];
-        h.update(&data);
-        hash.copy_from_slice(h.finalize().as_slice());
-        digest.copy_from_slice(&hash[..32]);
-        let digest = Digest(digest);
-
-        let (sender, receiver) = oneshot::channel();
-        signing_channel
-            .send((digest.clone(), sender))
-            .await
-            .map_err(|e| DagError::ChannelError {
-                error: format!("{}", e),
-            })?;
-        match receiver.await {
-            Ok(signature) => Ok(Self {
-                data,
-                digest,
-                signature,
-            }),
-            Err(e) => dag_bail!(DagError::ChannelError {
-                error: format!("{}", e)
-            }),
-        }
+impl fmt::Display for Header {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        write!(f, "B{}({})", self.round, self.author)
     }
+}
 
-    // Make a signed blockheader directly with a secret key.
-    // For debug / testing only.
-    pub fn debug_new(header: BlockHeader, secret_key: &SecretKey) -> Result<Self, DagError> {
-        let data: Vec<u8> = bincode::serialize(&header)?;
-        let digest = hash_vec(&data);
-        let signature = Signature::new(&digest, secret_key);
-        Ok(Self {
-            data,
-            digest,
-            signature,
-        })
-    }
+#[derive(Clone, Serialize, Deserialize)]
+pub struct Vote {
+    pub id: Digest,
+    pub round: Round,
+    pub origin: PublicKey,
+    pub author: PublicKey,
+    pub signature: Signature,
+}
 
-    /// Verify the signature and return the deserialized block header.
-    pub fn check(&self, committee: &Committee) -> Result<BlockHeader, DagError> {
-        // Check the digest is good.
-        dag_ensure!(hash_vec(&self.data) == self.digest, DagError::BadDigest);
+impl Vote {
+    pub async fn new(
+        header: &Header,
```

Comment:
might be a bit excessive to add the full header (O(n)) inside a vote, can we have the hash?